packages/hardhat-core/src/internal/core/providers/accounts.ts

```diff
@@ -146,57 +160,59 @@ export class LocalAccountsProvider extends ProviderWrapperWithChainId {
     }
   }
 
-  private _getPrivateKeyForAddress(address: string): Buffer {
-    const pk = this._addressToPrivateKey.get(address.toLowerCase());
+  private _getPrivateKeyForAddress(address: Buffer): Buffer {
+    const { bufferToHex } = require("ethereumjs-util");
+    const pk = this._addressToPrivateKey.get(bufferToHex(address));
     if (pk === undefined) {
       throw new HardhatError(ERRORS.NETWORK.NOT_LOCAL_ACCOUNT, {
-        account: address,
+        account: bufferToHex(address),
       });
     }
 
     return pk;
   }
 
-  private async _getNonceAsQuantity(address: string): Promise<string> {
-    return (await this._wrappedProvider.request({
+  private async _getNonce(address: Buffer): Promise<BN> {
+    const { bufferToHex } = await import("ethereumjs-util");
+
+    const response = (await this._wrappedProvider.request({
       method: "eth_getTransactionCount",
-      params: [address, "pending"],
+      params: [bufferToHex(address), "pending"],
     })) as string;
+
+    return rpcQuantityToBN(response);
   }
 
   private async _getSignedTransaction(
-    jsonRpcTransactionData: JsonRpcTransactionData,
+    transactionRequest: RpcTransactionRequest,
     chainId: number,
     privateKey: Buffer
   ): Promise<Buffer> {
     const chains = await import("@ethereumjs/common/dist/chains");
 
     const { Transaction } = await import("@ethereumjs/tx");
-    let transaction: TransactionT;
 
     const { default: Common } = await import("@ethereumjs/common");
 
     const txData = {
-      ...jsonRpcTransactionData,
-      gasLimit: jsonRpcTransactionData.gas,
+      ...transactionRequest,
+      gasLimit: transactionRequest.gas,
     };
 
-    if (chains.chains.names[chainId] !== undefined) {
-      transaction = Transaction.fromTxData(txData, {
-        common: new Common({ chain: chainId }),
-      });
-    } else {
-      const common = Common.forCustomChain(
-        "mainnet",
-        {
-          chainId,
-          networkId: chainId,
-        },
-        "istanbul"
-      );
+    const common =
```

Comment:
This wasn't really necessary. Made this stylistic change to make printf-debugging it easier.